### PR TITLE
add checks on active power control creation

### PR DESCRIPTION
**Pull request by EtienneLt**

### CODE REVIEW

**Review by ghazwarhili — COMMENTED** [review]
We have to add the check only when modifying the equipment

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java [hunk 1]
```diff
@@ -71,6 +72,15 @@ private void checkActivePowerZeroOrBetweenMinAndMaxActivePowerGenerator(Generato
         );
     }
 
+    private void checkActivePowerRegulation(GeneratorModificationInfos modificationInfos, Generator generator, NetworkModificationException.Type exceptionType, String errorMessage) {
+//        ModificationUtils.getInstance().checkActivePowerControl(
+//            modificationInfos.getParticipate() != null ? modificationInfos.getParticipate().getValue() : null,
+//            modificationInfos.getDroop() != null ? modificationInfos.getDroop().getValue() : null,
+//            exceptionType,
+//            errorMessage
+//        );
```
[the remark this answers is not in the document]
Comment (EtienneLt) [reply src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java]: forgot to remove

File: src/main/java/org/gridsuite/modification/server/modifications/BatteryModification.java [hunk 2]
```diff
@@ -171,7 +172,8 @@ public static ReportNode modifyBatteryActivePowerControlAttributes(AttributeModi
                                                                        ReportNode subReportNodeSetpoints) {
         ActivePowerControl<Battery> activePowerControl = battery.getExtension(ActivePowerControl.class);
         ActivePowerControlAdder<Battery> activePowerControlAdder = battery.newExtension(ActivePowerControlAdder.class);
-        return ModificationUtils.getInstance().modifyActivePowerControlAttributes(activePowerControl, activePowerControlAdder, participate, droop, subReportNode, subReportNodeSetpoints);
+        return ModificationUtils.getInstance().modifyActivePowerControlAttributes(activePowerControl, activePowerControlAdder,
+            participate, droop, subReportNode, subReportNodeSetpoints, CREATE_BATTERY_ERROR, "Battery '" + battery.getId() + "' : ");
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/BatteryModification.java:176]: MODIFY_BATTERY_ERROR not CREATE_BATTERY_ERROR
and you can create const variable errorMessage for CREATE_BATTERY_ERROR, "Battery '" + battery.getId() + "' : "

File: src/main/java/org/gridsuite/modification/server/dto/byfilter/equipmentfield/GeneratorField.java [hunk 3]
```diff
@@ -98,7 +98,8 @@ public static void setNewValue(Generator generator, String generatorField, @NotN
                 ActivePowerControl<Generator> activePowerControl = generator.getExtension(ActivePowerControl.class);
                 ActivePowerControlAdder<Generator> activePowerControlAdder = generator.newExtension(ActivePowerControlAdder.class);
                 ModificationUtils.getInstance().modifyActivePowerControlAttributes(activePowerControl, activePowerControlAdder, null,
-                        new AttributeModification<>(Float.parseFloat(newValue), OperationType.SET), null, null);
+                        new AttributeModification<>(Float.parseFloat(newValue), OperationType.SET), null, null,
+                    MODIFY_GENERATOR_ERROR, "Generator '" + generator.getId() + "' : ");
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/dto/byfilter/equipmentfield/GeneratorField.java:102]: you can create variable errorMessage for "Generator '" + generator.getId() + "' : "

File: src/main/java/org/gridsuite/modification/server/dto/byfilter/equipmentfield/BatteryField.java [hunk 4]
```diff
@@ -64,7 +64,8 @@ public static void setNewValue(Battery battery, String batteryField, @NotNull St
                 ActivePowerControlAdder<Battery> activePowerControlAdder = battery.newExtension(ActivePowerControlAdder.class);
                 ModificationUtils.getInstance().modifyActivePowerControlAttributes(
                         activePowerControl, activePowerControlAdder, null,
-                        new AttributeModification<>(Float.parseFloat(newValue), OperationType.SET), null, null);
+                        new AttributeModification<>(Float.parseFloat(newValue), OperationType.SET), null,
+                    null, MODIFY_BATTERY_ERROR, "Battery '" + battery.getId() + "' : ");
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/dto/byfilter/equipmentfield/BatteryField.java:68]: create const variable errorMessage for "Generator '" + generator.getId() + "' : "

File: src/main/java/org/gridsuite/modification/server/modifications/BatteryModification.java [hunk 5]
```diff
@@ -25,6 +25,7 @@
 import java.util.Collection;
 import java.util.List;
 
+import static org.gridsuite.modification.server.NetworkModificationException.Type.CREATE_BATTERY_ERROR;
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/BatteryModification.java:28]: remove

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java [hunk 6]
```diff
@@ -210,7 +210,9 @@ private ReportNode modifyGeneratorActivePowerControlAttributes(GeneratorModifica
                                                                  Generator generator, ReportNode subReportNode, ReportNode subReportNodeSetpoints) {
         ActivePowerControl<Generator> activePowerControl = generator.getExtension(ActivePowerControl.class);
         ActivePowerControlAdder<Generator> activePowerControlAdder = generator.newExtension(ActivePowerControlAdder.class);
-        return ModificationUtils.getInstance().modifyActivePowerControlAttributes(activePowerControl, activePowerControlAdder, modificationInfos.getParticipate(), modificationInfos.getDroop(), subReportNode, subReportNodeSetpoints);
+        return ModificationUtils.getInstance().modifyActivePowerControlAttributes(activePowerControl, activePowerControlAdder,
+            modificationInfos.getParticipate(), modificationInfos.getDroop(), subReportNode, subReportNodeSetpoints,
+            MODIFY_GENERATOR_ERROR, "Generator '" + modificationInfos.getEquipmentId() + "' : ");
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java:215]: the same for errorMessage

File: src/main/java/org/gridsuite/modification/server/modifications/ModificationUtils.java [hunk 7]
```diff
@@ -1184,38 +1184,51 @@ private void modifyExistingActivePowerControl(ActivePowerControl<?> activePowerC
...
-        adder.withParticipate(participate);
-        if (participateInfo != null && reports != null) {
-            reports.add(buildModificationReport(null, participate, "Participate"));
+                                             List<ReportNode> reports,
+                                             NetworkModificationException.Type exceptionType,
+                                             String errorMessage) {
+        Boolean participate = participateInfo == null ? null : participateInfo.getValue();
+        Float droop = droopInfo == null ? null : droopInfo.getValue();
+        checkActivePowerControl(participate, droop, exceptionType, errorMessage);
+        if (participate != null && droop != null) {
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/ModificationUtils.java:1193]:
```suggestion
        Boolean participate = Optional.ofNullable(participateInfo).map(AttributeModification::getValue).orElse(null);
        Float droop = Optional.ofNullable(droopInfo).map(AttributeModification::getValue).orElse(null);
        checkActivePowerControl(participate, droop, exceptionType, errorMessage);
        if (participate != null && droop != null) {
            adder.withParticipate(participate)
                    .withDroop(droop)
                    .add();
            if (reports != null) {
                reports.add(buildModificationReport(null, participate, "Participate"));
                reports.add(buildModificationReport(Double.NaN, droop, "Droop"));
            }
        }
    }
```

File: src/main/java/org/gridsuite/modification/server/modifications/ModificationUtils.java [hunk 8]
```diff
@@ -1184,38 +1184,51 @@ private void modifyExistingActivePowerControl(ActivePowerControl<?> activePowerC
...
         }
-        double droop = droopInfo != null ? droopInfo.getValue() : Double.NaN;
-        adder.withDroop(droop);
-        if (droopInfo != null && reports != null) {
-            reports.add(buildModificationReport(Double.NaN, droop, "Droop"));
+    }
+
+    public void checkActivePowerControl(Boolean participate, Float droop, NetworkModificationException.Type exceptionType, String errorMessage) {
+        if (Boolean.TRUE.equals(participate) && droop == null) {
+            throw new NetworkModificationException(exceptionType, errorMessage + "for creation if participate is true, droop can not be null");
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/ModificationUtils.java:1208]:
```suggestion
            if (Boolean.TRUE.equals(participate) && droop == null) {
            throw new NetworkModificationException(
                    exceptionType,
                    String.format("%s for creation: if participate is true, droop cannot be null", errorMessage)
            );
        }
```

**Review by ghazwarhili — CHANGES_REQUESTED** [review]

**Review by ghazwarhili — APPROVED** [review]

**Pull request merged**